src/ng/directive/ngRepeat.js

```diff
@@ -425,126 +419,124 @@ var ngRepeatDirective = ['$parse', '$animate', '$compile', function($parse, $ani
-              });
+
+            if (nextBlockMap[trackById]) {
+              // if collision detected, throw an error
               throw ngRepeatMinErr('dupes',
-                  'Duplicates in a repeater are not allowed. Use \'track by\' expression to specify unique keys. Repeater: {0}, Duplicate key: {1}, Duplicate value: {2}',
-                  expression, trackById, value);
-            } else {
-              // new never before seen block
-              nextBlockOrder[index] = {id: trackById, scope: undefined, clone: undefined};
-              nextBlockMap[trackById] = true;
+                'Duplicates in a repeater are not allowed. Use \'track by\' expression to specify unique keys. Repeater: {0}, Duplicate key: {1}, Duplicate value: {2}',
+                expression, trackById, value);
             }
-          }
 
-          // remove leftover items
-          for (var blockKey in lastBlockMap) {
-            block = lastBlockMap[blockKey];
-            elementsToRemove = getBlockNodes(block.clone);
-            $animate.leave(elementsToRemove);
-            if (elementsToRemove[0].parentNode) {
-              // if the element was not removed yet because of pending animation, mark it as deleted
-              // so that we can ignore it later
-              for (index = 0, length = elementsToRemove.length; index < length; index++) {
-                elementsToRemove[index][NG_REMOVED] = true;
-              }
-            }
-            block.scope.$destroy();
+            nextBlockMap[trackById] = {id: trackById, clone: undefined, scope: undefined, index: index};
+            nextBlockOrder[index] = trackById;
           }
 
-          // we are not using forEach for perf reasons (trying to avoid #call)
-          for (index = 0; index < collectionLength; index++) {
-            key = (collection === collectionKeys) ? index : collectionKeys[index];
-            value = collection[key];
-            block = nextBlockOrder[index];
+          // setup lastBlockOrder, used to determine if block moved
+          for (key in lastBlockMap) {
+            lastBlockOrder.push(key);
+          }
 
-            if (block.scope) {
-              // if we have already seen this object, then we need to reuse the
-              // associated scope/element
+          for (index = 0; index < nextLength; index++) {
+            key = collectionIsLikeArray ? index : collectionKeys[index];
+            nextKey = nextBlockOrder[index];
 
-              nextNode = previousNode;
+            if (lastBlockMap[nextKey]) {
+              // we have already seen this object and need to reuse the associated scope/element
+              block = lastBlockMap[nextKey];
 
-              // skip nodes that are already pending removal via leave animation
-              do {
-                nextNode = nextNode.nextSibling;
-              } while (nextNode && nextNode[NG_REMOVED]);
+              // move
```

Comment:
The logic of when to move and where to move is a key change in this pull request.